### ISSUE TITLE
A test case for diffing line 246

### ISSUE THREAD

**Issue by WebReflection** [snabbdom/snabbdom#348]
First of all, thanks for this project, it's been the key to port a reasonably fast DOM diffing algorithm to "[_vanilla_ DOM](https://github.com/WebReflection/domdiff#domdiff)" land.

Since I love to test-cover all branches of my code, I wonder if you could provide a basic example for [line 246](https://github.com/snabbdom/snabbdom/blob/master/src/snabbdom.ts#L246) which is never reached in any of my attempts.

Baiscally ...

```js
// this is is fine
if (oldStartIdx > oldEndIdx) {
  before = newCh[newEndIdx+1] == null ? null : newCh[newEndIdx+1].elm;
  addVnodes(parentElm, before, newCh, newStartIdx, newEndIdx, insertedVnodeQueue);
}
// this is always the counter case 
else if (newStartIdx > newEndIdx) {
  removeVnodes(parentElm, oldCh, oldStartIdx, oldEndIdx);
}
// while an else here would never be reached
```

Thanks in advance for any sort of clarification.

**Comment by caridy** [comment]
In theory, diffing a children collection that is shorting, then with another that is larger that the previous children collection should trigger both branches, e.g.:

```
var vnode1 = h('div', []);
var vnode2 = h('div', [h('p')]);
var vnode3 = h('div', []);
patch(elm, vnode1);
patch(vnode1, vnode2);
patch(vnode2, vnode3);
```

**Comment by WebReflection** [comment]
in that case I have `newStartIdx` at `0` and `newEndIdx` at `-1` because of [this line](https://github.com/snabbdom/snabbdom/blob/master/src/snabbdom.ts#L187) where the length of old or new list will be negative when the children list is empty.

**Comment by caridy** [comment]
Oh, this is a `while` statement, we should probably add more than empty children arrays there, instead have at least more than 1, so it goes thru the first pass, and then the rest should be outside, maybe:

```
var vnode1 = h('div', [h('p')]);
var vnode2 = h('div', [h('p'), h('h1'), h('h2')]);
var vnode3 = h('div', [h('p')]);
patch(elm, vnode1);
patch(vnode1, vnode2);
patch(vnode2, vnode3);
```

**Comment by WebReflection** [comment]
I'm afraid the moment you reach this point is where you exit the while loop and find futureStart 1 and futureEnd 0

https://github.com/snabbdom/snabbdom/blob/master/src/snabbdom.ts#L227-L240

I have dropped that check (and few others) and everything seems to be fine but I usually believe if a check was there maybe there is a weird case I'm not covering ... but I swear I've tried them all

**Comment by caridy** [comment]
Ok, I can run a big app with thousands of tests around this, and will report back.

**Comment by paldepind** [comment]
> First of all, thanks for this project, it's been the key to port a reasonably fast DOM diffing algorithm to "vanilla DOM" land.

I'm happy to hear that 😄 It's clear that you've been looking closely at the code. That is nice. Then we get good questions like this one 👍 

I just took a quick look at this. And as far as I can see the line you're talking about should never be reached.

We have this code above

```ts
while (oldStartIdx <= oldEndIdx && oldStartIdx <= oldEndIdxoldStartIdx <= oldEndIdx) {
  ...
}
```

So when we exit this `while`-loop we know that `oldStartIdx <= oldEndIdx && newStartIdx <= newEndIdx` no longer holds. From that we can deduce.

```
   !(oldStartIdx <= oldEndIdx && newStartIdx <= newEndIdx)
=> !(oldStartIdx <= oldEndIdx) || !(newStartIdx <= newEndIdx)
=> (oldStartIdx > oldEndIdx) || (newStartIdx > newEndIdx)
```

In other words we know exactly that either `oldStartIdx > oldEndIdx` or `newStartIdx > newEndIdx` must be true—otherwise we'd still be in the `while`-loop above.

As you probably know the `while`-loop does the diffing by working from the ends and inwards on both the old and new list of vnodes. Along the way it compares what it finds. At some point it stops, either because one of the lists are "exhausted" (i.e. no more work to do on them) or because both lists are exhausted. If both the lists are exhausted or if the old list is exhausted then the `if` branch will be taken and otherwise the `else if` branch will be taken. So it makes sense that the missing `else` branch can never happen.

Again, I looked at this slightly hastily so I may be wrong. But it seems to me that we could do

```diff
    if (oldStartIdx > oldEndIdx) {
      before = newCh[newEndIdx+1] == null ? null : newCh[newEndIdx+1].elm;
      addVnodes(parentElm, before, newCh, newStartIdx, newEndIdx, insertedVnodeQueue);
-    } else if (newStartIdx > newEndIdx) {
+    } else {
      removeVnodes(parentElm, oldCh, oldStartIdx, oldEndIdx);
    }
```

**Comment by WebReflection** [comment]
yup, that's what I'm doing and it works just fine indeed

**Comment by paldepind** [comment]
Btw, the Snabbdom test suite never reaches the `else` case either. We also have a [test case](https://github.com/snabbdom/snabbdom/blob/master/test/core.js#L557) that tests the diffing algorithm with randomly shuffled vnodes. I tried increasing the sample count and running it several times—still without reaching the line.

**Comment by WebReflection** [comment]
if it could help, I'm code covering 100% of domdiff code via this file:
https://github.com/WebReflection/domdiff/blob/master/test/test.js

which you can verify in your browser console via this page:
https://webreflection.github.io/domdiff/test/

**Comment by caridy** [comment]
@WebReflection those tests seems to be very exhaustive, any chance that you can send a PR for snabbdom to run those exact tests in the future?

**Comment by caridy** [comment]
Ok, I went into snabbdom (not the latest, but maybe a couple of months old version of it since we haven't upgraded recently), and add a throw in the `else` branch:

![image](https://user-images.githubusercontent.com/38969/33727038-e218b03c-db44-11e7-9f6b-20d55ef8aebc.png)

and I got exactly 3 failures:

```
Summary of all failing tests
 FAIL  src/framework/__tests__/vm-test.ts
  ● vm › integration › should assign bigger idx on reinsertion, including children idx

    ufff, something is wrong with issue #348

      at updateChildren (src/3rdparty/snabbdom/snabbdom.ts:230:19)
      at patchVnode (src/3rdparty/snabbdom/snabbdom.ts:256:17)
      at Object.patch (src/3rdparty/snabbdom/snabbdom.ts:284:13)

 FAIL  src/framework/__tests__/invoker-test.ts
  ● invoker › integration › should invoke disconnectedCallback() in a child after disconnectedCallback() on parent

    ufff, something is wrong with issue #348

      at updateChildren (src/3rdparty/snabbdom/snabbdom.ts:230:19)
      at patchVnode (src/3rdparty/snabbdom/snabbdom.ts:256:17)
      at Object.patch (src/3rdparty/snabbdom/snabbdom.ts:284:13)

  ● invoker › integration › should invoke disconnectedCallback() after it was removed from the dom

    ufff, something is wrong with issue #348

      at updateChildren (src/3rdparty/snabbdom/snabbdom.ts:230:19)
      at patchVnode (src/3rdparty/snabbdom/snabbdom.ts:256:17)
      at Object.patch (src/3rdparty/snabbdom/snabbdom.ts:284:13)
```

I will investigate more today, but it seems that either our tests are doing something funky, or we are not fully understanding this.

**Comment by WebReflection** [comment]
@caridy most of the tests *where* a re-adapted replica of snabbdom tests.

About your findings though , I’d love to understand if there’s a single use case where there’s nothing to do at the end and the inexistent else would be reached.

Thanks

**Comment by caridy** [comment]
You can look at some tests that are failing:

https://github.com/snabbdom/snabbdom/blob/0273fcf05b8134a12cdecff334f8175513d5a654/test/core.js#L874-L898

It fails on `assert.equal(1, result.length);`. The interesting case there is that it uses the remove callback mechanism, which I'm definitely not using in my big app.

**Comment by paldepind** [comment]
@caridy If I am understanding correctly you are adding the `throw` in the wrong place.

```
if (oldStartIdx > oldEndIdx) {
  before = newCh[newEndIdx+1] == null ? null : newCh[newEndIdx+1].elm;
  addVnodes(parentElm, before, newCh, newStartIdx, newEndIdx, insertedVnodeQueue);
} else if (newStartIdx > newEndIdx) {
  removeVnodes(parentElm, oldCh, oldStartIdx, oldEndIdx);
} else {
  throw new Error("This will never throw");
}
```

Both the `oldStartIdx > oldEndIdx` and the `newStartIdx > newEndIdx` branch does something useful. But the last branch can be converted from an `else if` to just an `else` since the check will always be `true` anyway.

**Comment by thysultan** [comment]
As far as coverage is concerned i think it might be safe put a Istanbul ignore comment to ignore the implied else branch. It will still hit but at that point i don't think there is anything useful left to do since from the looks of it, it has already been done in the previous [branch](https://github.com/snabbdom/snabbdom/blob/master/src/snabbdom.ts#L234-L238) that moves elements.

**Comment by thysultan** [comment]
That said, i'm not sure what branches it should take when you remove two nodes that are not adjacent to each other, then try to shuffle the list.

**Comment by caridy** [comment]
jaja! I was reading this whole thread backward! I now understand the original question, and yes, I think ignoring that branch from the coverage should be fine.

**Comment by WebReflection** [comment]
Yes @paldepind got what I meant.

You can’t remove the else if but safely convert into an else, unless I’m still missing something.

@thysultan adding an ignore for something that logically could never happen is not my style. I’ve opened this to understand what is the reason for an apparently always true else-if condition.

**Comment by thysultan** [comment]
Both the `else if` and `if` should not be true if you don't remove/add items, i.e just move elements around, so the implied else branch might hit in some cases but this is assuming `removeNodes` doesn't have a branch to handle this, so i might be wrong.

**Comment by WebReflection** [comment]
As previously explained, once out of that while one of those cases will always be true.

Use a code coverage tool to demonstrate the case you talk about ‘cause in every test of mine it never happens.

Both if and else-if are a guard for something else that might, true that, never happen.

Yet if last if is not true and we’re out of the loop that keeps incrementing indexes and reducing lengths of course last else if is always true.

However, if you find that shuffling case you’re mentioning both my code and snabbdom code would take advantage in covering such case.

**Comment by WebReflection** [comment]
To re-cap, @paldepind nailed the issue:

```
   !(oldStartIdx <= oldEndIdx && newStartIdx <= newEndIdx)
=> !(oldStartIdx <= oldEndIdx) || !(newStartIdx <= newEndIdx)
=> (oldStartIdx > oldEndIdx) || (newStartIdx > newEndIdx)
```

That makes last else-if like an else if(true) ... apologieses I’m on my phone and can’t write any better explanation without a laptop but I hope it’s clear what’s the redundant bit that puzzled me and my code-coverage related tests.

**Comment by thysultan** [comment]
These are the tests cases i use for DIO [Reconcile.spec.js](https://github.com/thysultan/dio.js/blob/master/test/Reconcile.spec.js)

It might be safe for snabbom to turn the `else if` to an `else` if it is always true but i think this is only because [snabbdom.ts#L236](https://github.com/snabbdom/snabbdom/blob/master/src/snabbdom.ts#L236) and [snabbdom.ts#L236](https://github.com/snabbdom/snabbdom/blob/master/src/snabbdom.ts#L161) make sure that moved nodes are never removed.

If domdiff implements a similar branch to safe guard against such cases then it is probably safe too.

**Comment by WebReflection** [comment]
domdiff uses exact same logic and indeed moves around nodes as well, that’s why is already shipping with last else/if converted as else.

**Comment by caridy** [comment]
Removing the `if` from the `else` is probably fine from the perspective of the logic, but not from the performance perspective. In fact I will like to strengthening that condition to avoid calling `removeNodes` all together if `oldStartIdx > oldEndIdx` because `removeNodes()` method is more expensive. There are other few places where something similar happen (e.g.: https://github.com/snabbdom/snabbdom/blob/master/src/snabbdom.ts#L245).

**Comment by WebReflection** [comment]
I’m not sure I’m following. Now you have an extra if that’s not necessary, since the else would be enough. That means wasted performance , right?

I

**Comment by blikblum** [comment]
> In fact I will like to strengthening that condition to avoid calling removeNodes all together if oldStartIdx > oldEndIdx because removeNodes() method is more expensive

The check is already done inside `removeNodes` which becomes a noop in this situation, so i don't see perf overhead. On the other side, adding the check in the caller would add redundant code

**Comment by caridy** [comment]
> The check is already done inside removeNodes which becomes a noop in this situation, so i don't see perf overhead. On the other side, adding the check in the caller would add redundant code

@diervo can you help with the perf analysis here? Usually, making a function call is more expensive that a condition. We can perf this and decide.

**Comment by paldepind** [comment]
@caridy 

> Usually, making a function call is more expensive than a condition. We can perf this and decide.

Yes, as long as the calls to `removeVnodes` and `addVnodes` is not inlined it will be more expensive than a condition. Removing the `if` from the `else` will not change how often these functions are called.

I do agree that there seems to be room for improvement in the code. What is the most often occurring case after the while-loop? I would argue that that is probably the case when nothing actually needs to be done. That happens both when there are no changes at all and when only the order of children has changed. In this case, we currently take the `if` branch and end up calling `addVnodes`. `addVnodes` then does nothing in its for-loop.

So, I agree that tightening the checks to avoid the function calls may be the best thing to do. Here is one way we could do that:

```js
if (oldStartIdx > oldEndIdx && newStartIdx <= newEndIdx) {
  before = newCh[newEndIdx+1] == null ? null : newCh[newEndIdx+1].elm;
  addVnodes(parentElm, before, newCh, newStartIdx, newEndIdx, insertedVnodeQueue);
} else if (oldStartIdx <= oldEndIdx) {
  removeVnodes(parentElm, oldCh, oldStartIdx, oldEndIdx);
}
```

The above code is untested. I've added and extra check to the first branch so that `addVnodes` is only called when it's for-loop will actually enter its body. In the else-branch, I've removed the current check (since we know it does nothing) and added a new check (the one you mentioned @caridy) that ensures that there are elements that needs removal.

Another approach might be to do wrap the if in another if.

```js
if (oldStartIdx <= oldEndIdx || newStartIdx <= newEndIdx) {
  if (oldStartIdx > oldEndIdx) {
    before = newCh[newEndIdx+1] == null ? null : newCh[newEndIdx+1].elm;
    addVnodes(parentElm, before, newCh, newStartIdx, newEndIdx, insertedVnodeQueue);
  } else {
    removeVnodes(parentElm, oldCh, oldStartIdx, oldEndIdx);
  }
}
```

This may be faster since it creates a fast path for the common case (when there is nothing to do after the while-loop).

**Comment by caridy** [comment]
I prefer the second approach!

**Comment by WebReflection** [comment]
FWIW latest domdiff ships with that second if guard and coverage is happy and also I am

**Comment by caridy** [comment]
Excellent! Second `if` also works fine for all tests, and some others app that I tested with. Here is the PR: https://github.com/snabbdom/snabbdom/pull/351

**Comment by caridy** [comment]
fixed by #351
